File: Robot Control Course/Chapter9/Lesson3/Chapter9_Lesson3.m

function [X, U] = ilqr_pendulum_matlab(x0, N, maxIter)
  m = 1.0; l = 1.0; g = 9.81; dt = 0.01;
  Q = diag([10, 1]);
  R = 0.1;
  P_N = diag([50, 5]);

  X = zeros(2, N+1);
  U = zeros(1, N);
  X(:,1) = x0(:);
  xref = zeros(2, N+1);

  % initial rollout
  for k = 1:N
    X(:,k+1) = dynamics(X(:,k), U(k), m, l, g, dt);
  end

  for it = 1:maxIter
    Vx = P_N * (X(:,N+1) - xref(:,N+1));
    Vxx = P_N;

    K = zeros(1, 2, N);
    k_ff = zeros(1, N);
    lambda = 1e-6;

    for k = N:-1:1
      xk = X(:,k);
      uk = U(k);
      xr = xref(:,k);

      [fx, fu] = linearize_dynamics(xk, uk, m, l, g, dt);

      dx = xk - xr;
      lx = Q * dx;
      lu = R * uk;
      lxx = Q;
      luu = R;

      Qx = lx + fx' * Vx;
      Qu = lu + fu' * Vx;
      Qxx = lxx + fx' * Vxx * fx;
      Quu = luu + fu' * Vxx * fu + lambda;
      Qux = fu' * Vxx * fx;

      k_local = -Quu \ Qu;
      K_local = -Quu \ Qux;

      k_ff(k) = k_local;
      K(1,:,k) = K_local;

      Vx = Qx + K_local' * Quu * k_local + K_local' * Qu + Qux' * k_local;
      Vxx = Qxx + K_local' * Quu * K_local + K_local' * Qux + Qux' * K_local;
      Vxx = 0.5 * (Vxx + Vxx');
    end

    % forward pass
    Xnew = zeros(size(X));
    Unew = zeros(size(U));
    Xnew(:,1) = x0(:);

    alpha_list = [1.0, 0.5, 0.25, 0.1];
    Jold = cost_total(X, U, xref, Q, R, P_N, N);

    improved = false;
    for a = alpha_list
      for k = 1:N
        du = a * k_ff(k) + squeeze(K(1,:,k)) * (Xnew(:,k) - X(:,k));
        Unew(k) = U(k) + du;
        Xnew(:,k+1) = dynamics(Xnew(:,k), Unew(k), m, l, g, dt);
      end
      Jnew = cost_total(Xnew, Unew, xref, Q, R, P_N, N);
      if Jnew < Jold
        X = Xnew; U = Unew;
        improved = true;
        break;
      end
    end

    if ~improved
      break;
    end
  end
end

function xnext = dynamics(x, u, m, l, g, dt)
  theta = x(1); dtheta = x(2);
  ddtheta = (-g / l) * sin(theta) + u / (m * l^2);
  xnext = [theta + dt * dtheta;
           dtheta + dt * ddtheta];
end

function [fx, fu] = linearize_dynamics(x, u, m, l, g, dt)
  theta = x(1);
  ddtheta_dtheta = (-g / l) * cos(theta);
  ddtheta_du = 1.0 / (m * l^2);

  fx = eye(2);
  fx(1,2) = dt;
  fx(2,1) = dt * ddtheta_dtheta;
  fx(2,2) = 1.0;

  fu = [0; dt * ddtheta_du];
end

function J = cost_total(X, U, xref, Q, R, P_N, N)
  J = 0;
  for k = 1:N
    dx = X(:,k) - xref(:,k);
    J = J + 0.5 * dx' * Q * dx + 0.5 * R * U(k)^2;
  end
  dxN = X(:,N+1) - xref(:,N+1);
  J = J + 0.5 * dxN' * P_N * dxN;
end

File: Tutorials/Codes/Robot_Control_Course/Chapter9/Lesson3/Chapter9_Lesson3.py

import numpy as np

# Pendulum parameters
m = 1.0
l = 1.0
g = 9.81
dt = 0.01

def dynamics(x, u):
    """Discrete-time pendulum dynamics x_{k+1} = f(x_k, u_k)."""
    theta, dtheta = x
    # continuous-time acceleration
    ddtheta = (-g / l) * np.sin(theta) + u / (m * l**2)
    # Euler integration
    theta_next = theta + dt * dtheta
    dtheta_next = dtheta + dt * ddtheta
    return np.array([theta_next, dtheta_next])

def linearize_dynamics(x, u):
    """Return f_x (2x2) and f_u (2x1) at (x, u)."""
    theta, dtheta = x
    # partial derivatives
    ddtheta_dtheta = (-g / l) * np.cos(theta)
    ddtheta_du = 1.0 / (m * l**2)

    fx = np.eye(2)
    fx[0, 1] = dt  # d theta_next / d dtheta
    fx[1, 0] = dt * ddtheta_dtheta
    fx[1, 1] = 1.0  # d dtheta_next / d dtheta

    fu = np.zeros((2, 1))
    fu[1, 0] = dt * ddtheta_du
    return fx, fu

# Cost matrices
Q = np.diag([10.0, 1.0])
R = np.array([[0.1]])
P_N = np.diag([50.0, 5.0])

def stage_cost(x, u, x_ref):
    dx = x - x_ref
    return 0.5 * dx.T @ Q @ dx + 0.5 * u.T @ R @ u

def terminal_cost(x, x_ref):
    dx = x - x_ref
    return 0.5 * dx.T @ P_N @ dx

def ilqr(x0, x_ref_traj, N, max_iter=50):
    # Initial (zero) control sequence
    n = x0.size
    m = 1
    U = np.zeros((N, m))
    X = np.zeros((N + 1, n))
    X[0] = x0

    # Rollout to initialize X
    for k in range(N):
        X[k + 1] = dynamics(X[k], U[k])

    for it in range(max_iter):
        # Backward pass
        Vx = P_N @ (X[N] - x_ref_traj[N])
        Vxx = P_N.copy()

        K = np.zeros((N, m, n))
        k_feedforward = np.zeros((N, m))

        regularization = 1e-6

        for k in range(N - 1, -1, -1):
            xk = X[k]
            uk = U[k]
            x_ref = x_ref_traj[k]

            fx, fu = linearize_dynamics(xk, uk)

            # Cost derivatives (quadratic tracking cost)
            dx = xk - x_ref
            lx = Q @ dx
            lu = R @ uk
            lxx = Q
            luu = R
            lux = np.zeros((m, n))

            # Q derivatives for iLQR
            Qx = lx + fx.T @ Vx
            Qu = lu + fu.T @ Vx
            Qxx = lxx + fx.T @ Vxx @ fx
            Quu = luu + fu.T @ Vxx @ fu + regularization * np.eye(m)
            Qux = lux + fu.T @ Vxx @ fx

            # Compute gains
            Quu_inv = np.linalg.inv(Quu)
            k_ff = -Quu_inv @ Qu
            K_fb = -Quu_inv @ Qux

            k_feedforward[k] = k_ff.ravel()
            K[k] = K_fb

            # Update value function
            Vx = Qx + K_fb.T @ Quu @ k_ff + K_fb.T @ Qu + Qux.T @ k_ff
            Vxx = Qxx + K_fb.T @ Quu @ K_fb + K_fb.T @ Qux + Qux.T @ K_fb
            # Symmetrize for numerical stability
            Vxx = 0.5 * (Vxx + Vxx.T)

        # Forward pass with line search
        alpha_list = [1.0, 0.5, 0.25, 0.1]
        cost_old = (
            sum(stage_cost(X[k], U[k], x_ref_traj[k]) for k in range(N))
            + terminal_cost(X[N], x_ref_traj[N])
        )
        improved = False
        for alpha in alpha_list:
            X_new = np.zeros_like(X)
            U_new = np.zeros_like(U)
            X_new[0] = x0
            for k in range(N):
                du = alpha * k_feedforward[k] + K[k] @ (X_new[k] - X[k])
                U_new[k] = U[k] + du
                X_new[k + 1] = dynamics(X_new[k], U_new[k])
            cost_new = (
                sum(stage_cost(X_new[k], U_new[k], x_ref_traj[k]) for k in range(N))
                + terminal_cost(X_new[N], x_ref_traj[N])
            )
            if cost_new < cost_old:
                X, U = X_new, U_new
                improved = True
                break

        if not improved:
            # Could increase regularization here; for simplicity we just stop
            break

    return X, U

# Example usage: stabilize around upright theta = 0
N = 300
x0 = np.array([0.5, 0.0])  # initial angle 0.5 rad
x_ref_traj = np.zeros((N + 1, 2))
X_opt, U_opt = ilqr(x0, x_ref_traj, N)
